File: sorting/insertion-sort/javascript/insertion_sort.js
/**
 * Insertion Sort Algorithm (JavaScript)
 *
 * Sorts an array by building a sorted section one element at a time.
 * This version does NOT modify the original array (non-mutating).
 *
 * Time Complexity:
 *   Best Case:  O(n)     - when array is already sorted
 *   Average Case: O(n²)
 *   Worst Case: O(n²)    - when array is reverse sorted
 * Space Complexity: O(n) - due to copying the input
 */

/**
 * Sorts an array using Insertion Sort (non-mutating).
 *
 * @param {Array} arr - The array of comparable elements to be sorted.
 * @returns {Array} A new sorted array (original is not modified).
 */
function insertionSort(arr) {
    // Create a copy to avoid mutating the original array
    const sorted = [...arr]; // Equivalent to arr.slice() or arr.copy()
    const n = sorted.length;

    // If array has 0 or 1 element, it's already sorted
    if (n < 2) return sorted;

    // Traverse from the second element (index 1) to the end
    for (let i = 1; i < n; i++) {
        const key = sorted[i]; // Current element to position
        let j = i - 1;         // Index of last element in sorted portion

        // Move elements greater than key one position ahead
        while (j >= 0 && sorted[j] > key) {
            sorted[j + 1] = sorted[j];
            j--;
        }

        // Insert the key at its correct position
        sorted[j + 1] = key;
    }

    return sorted;
}

// Example Usage
const numbers = [12, 11, 13, 5, 6];
console.log("Original Array:", numbers); // Remains unchanged

const sortedNumbers = insertionSort(numbers);
console.log("Sorted Array:", sortedNumbers);

// Verify original is unchanged
console.log("Original After Sort:", numbers);

// Test other cases
console.log("Already Sorted:", insertionSort([1, 2, 3, 4, 5]));
console.log("Reverse Sorted:", insertionSort([5, 4, 3, 2, 1]));
console.log("Single Element:", insertionSort([42]));
console.log("Empty Array:", insertionSort([]));

File: sorting/insertion-sort/python/insertion_sort.py
"""
Insertion Sort Algorithm (Python) - Non-Mutating Version

Sorts an array by building a sorted section one element at a time.
This version does NOT modify the original list.

Time Complexity:
    Best Case:  O(n)     - when array is already sorted
    Average Case: O(n²)
    Worst Case: O(n²)    - when array is reverse sorted
Space Complexity: O(n)   - due to copying the input
"""


def insertion_sort(arr):
    """
    Sorts a list using the Insertion Sort algorithm.
    
    This version creates a copy of the input list, so the original
    remains unchanged.

    Args:
        arr (list): The list of comparable elements to be sorted.

    Returns:
        list: A new list that is sorted.
    """
    # Create a copy to avoid mutating the original array
    sorted_arr = arr.copy()
    n = len(sorted_arr)

    # If array has 0 or 1 element, it's already sorted
    if n < 2:
        return sorted_arr

    # Traverse from the second element (index 1) to the end
    for i in range(1, n):
        key = sorted_arr[i]  # Current element to be positioned
        j = i - 1            # Index of the last element in the sorted portion

        # Move elements greater than key one position ahead
        while j >= 0 and sorted_arr[j] > key:
            sorted_arr[j + 1] = sorted_arr[j]
            j -= 1

        # Insert the key at its correct position
        sorted_arr[j + 1] = key

    return sorted_arr


# Example Usage
if __name__ == "__main__":
    numbers = [12, 11, 13, 5, 6]
    print("Original Array:", numbers)  # Remains unchanged

    sorted_numbers = insertion_sort(numbers)  # Safe: original not modified
    print("Sorted Array:", sorted_numbers)

    # Verify original is unchanged
    print("Original After Sort:", numbers)

    # Test other cases
    print("Already Sorted:", insertion_sort([1, 2, 3, 4, 5]))
    print("Reverse Sorted:", insertion_sort([5, 4, 3, 2, 1]))
    print("Single Element:", insertion_sort([42]))
    print("Empty Array:", insertion_sort([]))